cartographer/common/thread_pool_test.cc

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/common/thread_pool.h"
+
+#include "cartographer/common/make_unique.h"
+#include "gtest/gtest.h"
+
+namespace cartographer {
+namespace common {
+namespace {
+
+class Receiver {
+ public:
+  void Receive(int number) {
+    Mutex::Locker locker(&mutex_);
+    received_numbers_.push_back(number);
+  }
+
+  void WaitForNumberSequence(const std::list<int>& expected_numbers) {
+    bool have_enough_numbers = false;
+    while (!have_enough_numbers) {
+      common::MutexLocker locker(&mutex_);
+      have_enough_numbers = locker.AwaitWithTimeout(
+          [this, &expected_numbers]() REQUIRES(mutex_) {
+            return (received_numbers_.size() >= expected_numbers.size());
+          },
+          common::FromSeconds(0.1));
+    }
+    EXPECT_EQ(expected_numbers, received_numbers_);
+  }
+
+  std::list<int> received_numbers_;
+  Mutex mutex_;
+};
+
+TEST(ThreadPoolTest, RunTask) {
+  ThreadPool pool(1);
+  Receiver receiver;
+  auto task = common::make_unique<Task>();
+  task->SetWorkItem([&receiver]() { receiver.Receive(1); });
+  pool.ScheduleWhenReady(std::move(task));
+  receiver.WaitForNumberSequence({1});
+}
+
+TEST(ThreadPoolTest, RunWithDependency) {
+  ThreadPool pool(2);
+  Receiver receiver;
+  auto task_2 = common::make_unique<Task>();
+  task_2->SetWorkItem([&receiver]() { receiver.Receive(2); });
+  auto task_1 = common::make_unique<Task>();
+  task_1->SetWorkItem([&receiver]() { receiver.Receive(1); });
+  auto weak_task_1 = pool.ScheduleWhenReady(std::move(task_1));
+  task_2->AddDependency(weak_task_1);
+  pool.ScheduleWhenReady(std::move(task_2));
+  receiver.WaitForNumberSequence({1, 2});
+}
+
+TEST(ThreadPoolTest, RunWithOutOfScopeDependency) {
+  ThreadPool pool(2);
+  Receiver receiver;
+  auto task_2 = common::make_unique<Task>();
+  task_2->SetWorkItem([&receiver]() { receiver.Receive(2); });
+  {
+    auto task_1 = common::make_unique<Task>();
+    task_1->SetWorkItem([&receiver]() { receiver.Receive(1); });
+    auto weak_task_1 = pool.ScheduleWhenReady(std::move(task_1));
+    task_2->AddDependency(weak_task_1);
+    // task_1 goes out of scope.
```

Comment:
I do not think that this comment is needed.

Reply:
Done